server/src/main/java/com/flipkart/varadhi/events/EntityEventDispatcher.java

```diff
@@ -0,0 +1,222 @@
+package com.flipkart.varadhi.events;
+
+import com.flipkart.varadhi.cluster.messages.ClusterMessage;
+import com.flipkart.varadhi.cluster.messages.ResponseMessage;
+import com.flipkart.varadhi.common.events.EntityEvent;
+import com.flipkart.varadhi.common.events.EntityEventListener;
+import com.flipkart.varadhi.common.utils.JsonMapper;
+import com.flipkart.varadhi.entities.MetaStoreEntity;
+import com.flipkart.varadhi.entities.Project;
+import com.flipkart.varadhi.entities.VaradhiTopic;
+import com.flipkart.varadhi.entities.auth.ResourceType;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A lean dispatcher for entity events to appropriate listeners.
+ * <p>
+ * The dispatcher routes entity events to registered listeners based on resource type,
+ * handles error cases gracefully, and provides detailed logging for operational visibility.
+ *
+ * @see EntityEvent
+ * @see EntityEventListener
+ * @see ResourceType
+ */
+@Slf4j
+public final class EntityEventDispatcher {
+
+    /**
+     * Map of resource types to their corresponding event listeners.
+     * This map is immutable after construction for thread safety.
+     */
+    private final Map<ResourceType, EntityEventListener<?>> listeners;
+
+    /**
+     * Set of resource types supported by this dispatcher.
+     * This set is derived from the listeners map and is immutable.
+     */
+    private final Set<ResourceType> supportedTypes;
+
+    /**
+     * Creates a new EntityEventDispatcher with the specified listeners.
+     *
+     * @param listeners a map of resource types to their corresponding event listeners
+     */
+    private EntityEventDispatcher(Map<ResourceType, EntityEventListener<?>> listeners) {
+        this.listeners = Map.copyOf(listeners);
+        this.supportedTypes = Set.copyOf(this.listeners.keySet());
+        log.info("Initialized EntityStateProcessor with supported types: {}", supportedTypes);
+    }
+
+    /**
+     * Processes an entity event from a cluster message.
+     * <p>
+     * This method extracts the entity event from the message, validates it,
+     * and routes it to the appropriate listener based on the resource type.
+     *
+     * @param message the cluster message containing the entity event
+     * @param <T>     the type of entity in the event, must extend MetaStoreEntity
+     * @return a future that completes with the response message
+     */
+    @SuppressWarnings ("unchecked")
+    public <T extends MetaStoreEntity> CompletableFuture<ResponseMessage> processEvent(ClusterMessage message) {
+        if (message == null) {
+            return createErrorResponse(null, "Message cannot be null", null);
+        }
```

Comment:
why are we putting defensive null checks everywhere?

There is no reason this method will be called with null message.